File: src/main/java/jadx/plugins/example/AliucordJADX.java
package jadx.plugins.example;

import java.util.Formatter;
import java.util.Locale;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import javax.swing.JOptionPane;

import jadx.api.metadata.ICodeNodeRef;
import jadx.api.plugins.JadxPlugin;
import jadx.api.plugins.JadxPluginContext;
import jadx.api.plugins.JadxPluginInfo;
import jadx.api.plugins.JadxPluginInfoBuilder;
import jadx.api.plugins.gui.JadxGuiContext;
import jadx.core.dex.instructions.args.ArgType;
import jadx.core.dex.nodes.FieldNode;
import jadx.core.dex.nodes.MethodNode;
import jadx.core.utils.exceptions.JadxRuntimeException;

public class AliucordJADX implements JadxPlugin {
	public static final String PLUGIN_ID = "aliucord-jadx";

	private final AliucordJADXOptions options = new AliucordJADXOptions();

	Function<ICodeNodeRef, Boolean> isHookActionEnabled = (node) -> node instanceof MethodNode;
	Function<ICodeNodeRef, Boolean> isFieldActionEnabled = (node) -> node instanceof FieldNode && options.getCodegenLanguage() == CodegenLanguage.KOTLIN;

	private static final String KOTLIN_CONSTRUCTOR_TEMPLATE = "patcher.%1$s<%2$s>(%3$s) { param -> \n // code \n }";
	private static final String KOTLIN_METHOD_TEMPLATE = "patcher.%1$s<%2$s>(%3$s%4$s) { param -> \n // code \n }";
	private static final String KOTLIN_FIELD_TEMPLATE = "var %1$s.exampleField by accessField<%2$s>(%3$s)";
	private static final String JAVA_CONSTRUCTOR_TEMPLATE = "patcher.patch(%2$s.getDeclaredConstructor(%3$s), new %1$s(param -> { \n // code \n }))";
	private static final String JAVA_METHOD_TEMPLATE = "patcher.patch(%2$s.class.getDeclaredMethod(\"%3$s\", %4$s), new %1$s(param -> { \n // code \n }))";

	private final Map<String, String> PRIMITIVE_TYPE_MAPPING = Map.of("int", "Int", "byte", "Byte", "short", "Short", "long", "Long", "float", "Float", "double", "Double", "char", "Char", "boolean", "Boolean");
	private final Map<String, String> BOXED_TYPE_MAPPING = Map.of("java.lang.Integer", "Int", "java.lang.Byte", "Byte", "java.lang.Short", "Short", "java.lang.Long", "Long", "java.lang.Float", "Float", "java.lang.Double", "Double", "java.lang.Character", "Char", "java.lang.Boolean", "Boolean", "java.lang.Object", "Any");

	@Override
	public JadxPluginInfo getPluginInfo() {
		return JadxPluginInfoBuilder.pluginId(PLUGIN_ID).name("Aliucord-JADX").description("Aliucord helper plugin").homepage("https://github.com/Canny1913/Aliucord-JADX").requiredJadxVersion("1.5.1, r2333").build();
	}

	@Override
	public void init(JadxPluginContext context) {
		context.registerOptions(options);
		var guiContext = context.getGuiContext();
		if (guiContext != null) {
			initGui(guiContext);
		}
	}

	public void initGui(JadxGuiContext context) {
		context.addPopupMenuAction(
				"Copy Aliucord hook snippet (before)",
				isHookActionEnabled, null,
				(nodeRef) -> snippetAction(context, nodeRef, HookType.BEFORE)
		);
		context.addPopupMenuAction(
				"Copy Aliucord hook snippet (after)",
				isHookActionEnabled, null,
				(nodeRef) -> snippetAction(context, nodeRef, HookType.AFTER)
		);
		context.addPopupMenuAction(
				"Copy Aliucord hook snippet (instead)",
				isHookActionEnabled, null,
				(nodeRef) -> snippetAction(context, nodeRef, HookType.INSTEAD)
		);
		context.addPopupMenuAction(
				"Copy Aliucord accessor snippet (field)",
				isFieldActionEnabled, null,
				(nodeRef) -> snippetAction(context, nodeRef, null)
		);
	}

	public void snippetAction(JadxGuiContext context, ICodeNodeRef nodeRef, HookType hookType) {
		try {
			var snippet = copySnippet(nodeRef, hookType);
			context.copyToClipboard(snippet);
		} catch (Exception e) {
			JOptionPane.showMessageDialog(context.getMainFrame(), e.getLocalizedMessage(), "Error", JOptionPane.ERROR_MESSAGE);
		}
	}

	public String copySnippet(ICodeNodeRef node, HookType hookType) {
		if (node instanceof MethodNode) {
			return copyMethodSnippet((MethodNode) node, hookType);
		} else if (node instanceof FieldNode) {
			return copyFieldSnippet((FieldNode) node);
		} else {
			throw new JadxRuntimeException("Unsupported node type.");
		}
	}


	public String copyMethodSnippet(MethodNode node, HookType hookType) {
		var language = options.getCodegenLanguage();
		var method = node.getMethodInfo();
		var args = node.getArgTypes().stream().map(this::fixType);
		var clazz = node.getDeclaringClass();
		var methodName = method.getName();
		var hookTypeStr = hookType.asString(language);
		String className;

		if (options.useFullName()) {
			className = clazz.getRawName();
		} else {
			className = clazz.getName();
		}
		if (language == CodegenLanguage.JAVA) {
			className = String.format("%s.class", className);
		}
		if (language == CodegenLanguage.KOTLIN) {
			if (className.contains("$")) className = String.format("`%s`", className);
			if (methodName.contains("$")) {
				methodName = String.format("$$\"%s\"", methodName);
			} else {
				methodName = String.format("\"%s\"", methodName);
			}
		}
		if (node.getAccessFlags().isAbstract()) {
			throw new JadxRuntimeException("Cannot create a snippet of abstract method.");
		}

		String constructorTemplate;
		String methodTemplate;
		if (language == CodegenLanguage.KOTLIN) {
			constructorTemplate = KOTLIN_CONSTRUCTOR_TEMPLATE;
			methodTemplate = KOTLIN_METHOD_TEMPLATE;
		} else {
			constructorTemplate = JAVA_CONSTRUCTOR_TEMPLATE;
			methodTemplate = JAVA_METHOD_TEMPLATE;
		}

		StringBuilder sb = new StringBuilder();
		Formatter formatter = new Formatter(sb, Locale.ROOT);
		if (method.isConstructor()) {
			formatter.format(constructorTemplate, hookTypeStr, className, args.collect(Collectors.joining(",\n", ",\n", "")));
		} else {
			formatter.format(methodTemplate, hookTypeStr, className, methodName, args.collect(Collectors.joining(",\n", ",\n", "")));
		}
		return sb.toString();
	}

	public String copyFieldSnippet(FieldNode node) {
		var language = options.getCodegenLanguage();

		var field = node.getFieldInfo();
		var fieldType = getFieldType(node.getType());
		var clazz = node.getDeclaringClass();
		var fieldName = field.getName();
		String className;
		if (options.useFullName()) {
			className = clazz.getRawName();
		} else {
			className = clazz.getName();
		}

		if (language == CodegenLanguage.KOTLIN) {
			if (className.contains("$")) className = String.format("`%s`", className);
			if (fieldName.contains("$")) {
				fieldName = String.format("$$\"%s\"", fieldName);
			} else {
				fieldName = String.format("\"%s\"", fieldName);
			}
		}
		if (node.getAccessFlags().isAbstract()) {
			throw new JadxRuntimeException("Cannot create a getter snippet of abstract field.");
		}
		StringBuilder sb = new StringBuilder();
		Formatter formatter = new Formatter(sb, Locale.ROOT);
		formatter.format(KOTLIN_FIELD_TEMPLATE, className, fieldType, fieldName);
		return sb.toString();
	}

	public String getFieldType(ArgType type) {
		var language = options.getCodegenLanguage();
		if (language == CodegenLanguage.KOTLIN) {
			if (type.canBeArray()) {
				var arrayElementType = type.getArrayElement();
				return String.format("Array<%s>", processTypeName(arrayElementType.toString(), false));
			}
			if (type.isGenericType() && type.isObject() && type.isTypeKnown()) {
				return processTypeName("java.lang.Object", false);
			}
			if (type.isGeneric() && type.isObject()) {
				var generics = type.getGenericTypes().stream().map((a) -> "*").collect(Collectors.joining(","));
				return String.format("%s<%s>", processTypeName(type.getObject(), false), generics);
			}
			if (type.isPrimitive()) {
				var pt = type.getPrimitiveType();
				return PRIMITIVE_TYPE_MAPPING.get(pt.toString());
			}
		}
		return processTypeName(type.toString(), false);
	}

	public String fixType(ArgType type) {
		var language = options.getCodegenLanguage();

		if (type.isGeneric() && type.isObject()) {
			if (type.getInnerType() != null) {
				return processTypeName(type.toString(), true);
			}
			return processTypeName(type.getObject(), true);
		}
		if (type.isGenericType() && type.isObject() && type.isTypeKnown()) {
			return processTypeName("java.lang.Object", true);
		}
		if (type.isPrimitive()) {
			if (language == CodegenLanguage.JAVA) {
				return String.format("%s.class", type);
			} else if (language == CodegenLanguage.KOTLIN) {
				var pt = type.getPrimitiveType();
				switch (pt) {
					case BOOLEAN:
						return "Boolean::class.javaPrimitiveType!!";
					case CHAR:
						return "Char::class.javaPrimitiveType!!";
					case BYTE:
						return "Byte::class.javaPrimitiveType!!";
					case SHORT:
						return "Short::class.javaPrimitiveType!!";
					case INT:
						return "Int::class.javaPrimitiveType!!";
					case FLOAT:
						return "Float::class.javaPrimitiveType!!";
					case LONG:
						return "Long::class.javaPrimitiveType!!";
					case DOUBLE:
						return "Double::class.javaPrimitiveType!!";
					case OBJECT:
						return "Any::class.java!!";
					case ARRAY:
						return "Array::class.java!!";
					case VOID:
						return "Void::class.javaPrimitiveType!!";
					default:
						throw new JadxRuntimeException("Unknown or null primitive type: " + type);
				}
			}
		}
		return processTypeName(type.toString(), true);
	}

	public String processTypeName(String type, boolean asClass) {
		var useFullName = options.useFullName();
		var language = options.getCodegenLanguage();
		String newType = type;
		if (!useFullName) {
			var lastIndex = newType.lastIndexOf(".");
			if (lastIndex != -1) newType = type.substring(lastIndex + 1);
		}
		if (asClass) {
			switch (language) {
				case KOTLIN:
					if (BOXED_TYPE_MAPPING.containsKey(type)) {
						return String.format("%s::class.javaObjectType", BOXED_TYPE_MAPPING.get(type));
					} else {
						return String.format("%s::class.java", newType);
					}
				case JAVA:
					return String.format("%s.class", newType);
			}
		}
		return newType;
	}
}

File: src/main/java/com/github/canny1913/AliucordJADX.kt
package com.github.canny1913

import jadx.api.metadata.ICodeNodeRef
import jadx.api.plugins.JadxPlugin
import jadx.api.plugins.JadxPluginContext
import jadx.api.plugins.JadxPluginInfo
import jadx.api.plugins.JadxPluginInfoBuilder
import jadx.api.plugins.gui.JadxGuiContext
import jadx.core.dex.instructions.args.ArgType
import jadx.core.dex.instructions.args.PrimitiveType
import jadx.core.dex.nodes.FieldNode
import jadx.core.dex.nodes.MethodNode
import jadx.core.utils.exceptions.JadxRuntimeException
import javax.swing.JOptionPane

class AliucordJADX : JadxPlugin {
	val options = AliucordJADXOptions()
	override fun getPluginInfo(): JadxPluginInfo? {
		return JadxPluginInfoBuilder.pluginId(PLUGIN_ID).name("Aliucord-JADX")
			.description("Aliucord helper plugin")
			.homepage("https://github.com/Canny1913/Aliucord-JADX")
			.requiredJadxVersion("1.5.1, r2333").build()
	}

	fun isHookActionEnabled(node: ICodeNodeRef) = node is MethodNode

	fun isFieldActionEnabled(node: ICodeNodeRef) =
		node is FieldNode && options.codegenLanguage.isKotlin()

	override fun init(context: JadxPluginContext?) {
		context?.registerOptions(options)
		val guiContext = context?.guiContext ?: return
		initGui(guiContext)
	}

	fun initGui(context: JadxGuiContext) {
		context.addPopupMenuAction(
			"Copy Aliucord hook snippet (before)", ::isHookActionEnabled, null
		) {
			snippetAction(context, it, HookType.BEFORE)
		}
		context.addPopupMenuAction(
			"Copy Aliucord hook snippet (after)", ::isHookActionEnabled, null
		) {
			snippetAction(context, it, HookType.AFTER)
		}
		context.addPopupMenuAction(
			"Copy Aliucord hook snippet (instead)", ::isHookActionEnabled, null
		) {
			snippetAction(context, it, HookType.INSTEAD)
		}
		context.addPopupMenuAction(
			"Copy Aliucord accessor snippet (field)", ::isFieldActionEnabled, null
		) {
			snippetAction(context, it)
		}
	}

	fun snippetAction(context: JadxGuiContext, node: ICodeNodeRef, hookType: HookType? = null) {
		try {
			val snippet = copySnippet(node, hookType)
			context.copyToClipboard(snippet)
		} catch (t: Throwable) {
			JOptionPane.showMessageDialog(
				context.mainFrame, t.localizedMessage, "Error", JOptionPane.ERROR_MESSAGE
			)

		}
	}

	fun copySnippet(node: ICodeNodeRef, hookType: HookType? = null): String {
		return when (node) {
			is MethodNode -> {
				copyMethodSnippet(node, hookType!!)
			}

			is FieldNode -> {
				copyFieldSnippet(node)
			}

			else -> {
				throw JadxRuntimeException("Unsupported node type.")
			}
		}
	}

	fun copyMethodSnippet(node: MethodNode, hookType: HookType): String {
		val language = options.codegenLanguage
		val method = node.methodInfo
		val args = node.argTypes.map(this::fixType)
		val clazz = node.declaringClass
		var methodName = method.name
		val hookTypeStr = hookType.asString(language)

		var className = if (options.useFullName) {
			clazz.rawName
		} else {
			clazz.name
		}
		if (language.isJava()) {
			className = "$className.class"
		}
		if (language.isKotlin()) {
			if (className.contains('$')) className = "`$className`"
			methodName = if (methodName.contains('$')) {
				"$$\"$methodName\""
			} else {
				"\"$methodName\""
			}
		}
		if (node.accessFlags.isAbstract) throw JadxRuntimeException("Cannot create a snippet of abstract method.")
		return if (method.isConstructor) {
			val formattedArgs = args.joinToString(",\n")
			if (options.codegenLanguage.isKotlin()) {
				"patcher.$hookTypeStr<$className>($formattedArgs) { param -> \n // code \n }"
			} else {
				"patcher.patch($className.getDeclaredConstructor($formattedArgs), new $hookTypeStr(param -> { \n // code \n }))"
			}
		} else {
			val formattedArgs = args.joinToString(",\n", ",\n")
			if (options.codegenLanguage.isKotlin()) {
				"patcher.$hookTypeStr<$className>($methodName$formattedArgs) { param -> \n // code \n }"
			} else {
				"patcher.patch($className.getDeclaredMethod($methodName$formattedArgs), new $hookTypeStr(param -> { \n // code \n }))"
			}
		}
	}

	fun copyFieldSnippet(node: FieldNode): String {
		val language = options.codegenLanguage
		val field = node.fieldInfo
		val fieldType = getFieldType(node.type)
		val clazz = node.declaringClass
		var fieldName = field.name

		var className = if (options.useFullName) {
			clazz.rawName
		} else {
			clazz.name
		}

		if (language.isKotlin()) {
			if (className.contains('$')) className = "`$className`"
			fieldName = if (fieldName.contains('$')) {
				"$$\"$fieldName\""
			} else {
				"\"$fieldName\""
			}
		}

		if (node.accessFlags.isAbstract) throw JadxRuntimeException("Cannot create a getter snippet of abstract field.")
		return "var $className.exampleField by accessField<$fieldType>($fieldName)"
	}

	fun getFieldType(type: ArgType): String {
		val language = options.codegenLanguage
		if (language.isKotlin()) {
			if (type.canBeArray()) {
				val arrayElementType = type.arrayElement
				val processedType = processTypeName(arrayElementType.toString(), false)
				return "Array<$processedType>"
			}
			if (type.isGenericType && type.isObject && type.isTypeKnown) {
				processTypeName("java.lang.Object", false)
			}
			if (type.isGeneric && type.isObject) {
				val generics = type.genericTypes.map { "*" }.joinToString { "," }
				val processedType = processTypeName(type.`object`, false)
				return "$processedType<$generics>"
			}
			if (type.isPrimitive) {
				val pt = type.primitiveType
				return PRIMITIVE_TYPE_MAPPING[pt.toString()]!!
			}
		}
		return processTypeName(type.toString(), false)
	}

	fun fixType(type: ArgType): String {
		val language = options.codegenLanguage
		if (type.isGeneric && type.isObject) {
			return if (type.innerType != null) {
				processTypeName(type.toString(), true)
			} else {
				processTypeName(type.`object`, true)
			}
		}
		if (type.isGenericType && type.isObject && type.isTypeKnown) {
			processTypeName("java.lang.Object", false)
		}
		if (type.isPrimitive) {
			if (language.isJava()) return "$type.class"
			val pt = type.primitiveType
			return when (pt) {
				PrimitiveType.BOOLEAN -> "Boolean::class.javaPrimitiveType!!"
				PrimitiveType.CHAR -> "Char::class.javaPrimitiveType!!"
				PrimitiveType.BYTE -> "Byte::class.javaPrimitiveType!!"
				PrimitiveType.SHORT -> "Short::class.javaPrimitiveType!!"
				PrimitiveType.INT -> "Int::class.javaPrimitiveType!!"
				PrimitiveType.FLOAT -> "Float::class.javaPrimitiveType!!"
				PrimitiveType.LONG -> "Long::class.javaPrimitiveType!!"
				PrimitiveType.DOUBLE -> "Double::class.javaPrimitiveType!!"
				PrimitiveType.OBJECT -> "Any::class.java!!"
				PrimitiveType.ARRAY -> "Array::class.java!!"
				PrimitiveType.VOID -> "Void::class.javaPrimitiveType!!"
				else -> throw JadxRuntimeException("Unknown or null primitive type: $type")
			}
		}
		return processTypeName(type.toString(), true)
	}

	fun processTypeName(type: String, asClass: Boolean): String {
		val useFullName = options.useFullName
		val language = options.codegenLanguage
		var newType = type
		if (!useFullName) {
			newType = type.substringAfterLast('.')
		}
		if (asClass) {
			if (language.isJava()) return "$newType.class"
			return if (type in BOXED_TYPE_MAPPING) {
				"$newType::class.javaObjectType"
			} else {
				"$newType:class.java"
			}
		} else {
			return newType
		}
	}

	internal companion object {
		const val PLUGIN_ID = "aliucord-jadx"

		private val PRIMITIVE_TYPE_MAPPING: Map<String, String> = mapOf(
			"int" to "Int",
			"byte" to "Byte",
			"short" to "Short",
			"long" to "Long",
			"float" to "Float",
			"double" to "Double",
			"char" to "Char",
			"boolean" to "Boolean"
		)
		private val BOXED_TYPE_MAPPING: Map<String, String> = mapOf(
			"java.lang.Integer" to "Int",
			"java.lang.Byte" to "Byte",
			"java.lang.Short" to "Short",
			"java.lang.Long" to "Long",
			"java.lang.Float" to "Float",
			"java.lang.Double" to "Double",
			"java.lang.Character" to "Char",
			"java.lang.Boolean" to "Boolean",
		)
	}
}

